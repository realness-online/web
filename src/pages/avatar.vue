<template>
  <section id="avatar" class="page">
    <header>
      <icon name=""></icon>
      <a @click='finished_viewing'>
        <icon name="finished"></icon>
      </a>
    </header>
    <icon v-if="working" name="working"></icon>
    <profile-as-figure v-else :person='person' :just_display_avatar="true"></profile-as-figure>
  </section>
</template>
<script>
  import icon from '@/components/icon'
  import profileAsFigure from '@/components/profile/as-figure'
  import profile_init from '@/mixins/profile_init'
  export default {
    mixins: [profile_init],
    components: {
      icon, profileAsFigure
    },
    methods: {
      finished_viewing() {
        const route = {
          path: sessionStorage.previous
        }
        this.$router.push(route)
      }
    }
  }
</script>
<style lang="stylus">
  @require '../style/variables'
  section#avatar
    animation-name: slideInLeft
    height: 80vh
    display: flex
    flex-direction: column
    justify-content: flex-start
    align-content: stretch
    figure.profile
      margin-top: base-line * 1.5
      padding: 0
      flex-grow: 1
      display: flex
      flex-direction: column
      justify-content: flex-start
      align-content: stretch
      & > svg
        animation-name: slideInDown
        border-radius: 100vw
        width: 100vw
        height:100vw
        max-height: 90vh
      & > figcaption
        display:none
</style>
