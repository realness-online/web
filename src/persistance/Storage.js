/** @typedef {import('@/types').Item_Id} Item_Id */

// https://developers.caffeina.com/object-composition-patterns-in-javascript-4853898bb9d0
import Local from '@/persistance/Local'
import Large from '@/persistance/Large'
import Cloud from '@/persistance/Cloud'
import Paged from '@/persistance/Paged'
import { current_user, upload } from '@/use/serverless'
import { as_type, as_filename } from '@/utils/itemid'
import { get } from 'idb-keyval'
import { prepare_upload_html } from '@/utils/upload_processor'

export default class Storage {
  /**
   * @param {Item_Id} itemid
   */
  constructor(itemid) {
    this.id = itemid
    this.type = as_type(itemid)
    this.metadata = { contentType: 'text/html' }
  }
}

/** @extends {Storage} */
export class Poster extends Large(Cloud(Storage)) {}

/** @extends {Storage} */
export class Me extends Cloud(Local(Storage)) {
  constructor() {
    super(localStorage.me)
  }
}

/** @extends {Storage} */
export class Relation extends Local(Storage) {
  constructor() {
    super(`${localStorage.me}/relations`)
  }
}

/** @extends {Storage} */
export class Statement extends Paged(Cloud(Local(Storage))) {
  constructor() {
    super(`${localStorage.me}/statements`)
  }
}

/** @extends {Storage} */
export class Event extends Paged(Cloud(Local(Storage))) {
  constructor() {
    super(`${localStorage.me}/events`)
  }
}

/** @extends {Storage} */
export class Offline extends Cloud(Storage) {
  async save() {
    const outerHTML = await get(this.id)
    if (outerHTML) await super.save({ outerHTML })
  }
}

/** @extends {Storage} */
export class History extends Cloud(Storage) {
  /**
   * @param {Item_Id} itemid
   */
  constructor(itemid) {
    super(itemid)
  }

  async save(items) {
    // on purpose doesn't call super.save
    if (!items) return false
    if (current_user.value && navigator.onLine) {
      const { compressed, metadata } = await prepare_upload_html(items)
      try {
        await upload(as_filename(this.id), compressed, metadata)
      } catch (e) {
        console.error(e)
        return false
      }
      return true
    }
    return false
  }
}
