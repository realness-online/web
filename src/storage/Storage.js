// https://developers.caffeina.com/object-composition-patterns-in-javascript-4853898bb9d0
import Item from '@/modules/item'
import Paged from '@/storage/Paged'
import Large from '@/storage/Large'
import Cloud from '@/storage/Cloud'

class Storage {
  constructor(type,
              selector = `[itemprop="/${type}"]`,
              filename = `${type}/index.html`,
              content_type = 'text/html') {
    this.type = type
    this.selector = selector
    this.filename = filename
    this.metadata = { 'contentType': content_type }
  }
  as_list() {
    return Item.get_items(localStorage.getItem(this.selector))
  }
  as_object() {
    return Item.get_first_item(localStorage.getItem(this.selector))
  }
  async save(items = document.querySelector(this.selector)) {
    if (!items) return
    localStorage.setItem(this.selector, items.outerHTML)
  }
}
class Person extends Cloud(Storage) {
  constructor() {
    super('person', '[itemid="person.html"]', 'person.html')
  }
}
class Post extends Paged(Cloud(Storage)) {
  constructor() {
    super('posts')
  }
}
class SVG extends Large(Cloud(Storage)) {}

export default Storage
export const relations_storage = new Storage('relations')
export const person_storage = new Person()
export const posts_storage = new Posts()
export const avatars_storage = new SVG('avatars')
export const posters_storage = new SVG('posters')
