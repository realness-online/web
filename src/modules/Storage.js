import Item from '@/modules/Item'
import * as firebase from 'firebase/app'
import 'firebase/auth'
import 'firebase/storage'

class Storage {
  static hydrate(item_as_string) {
    return document.createRange().createContextualFragment(item_as_string)
  }
  constructor(item_type, selector = `[itemtype="/${item_type}"]`,
    filename = `${item_type}.html`) {
    this.item_type = item_type
    this.selector = selector
    this.filename = filename
    this.metadata = {'contentType': 'text/html'}
  }
  from_storage() {
    let storage_string = localStorage.getItem(this.item_type)
    return Storage.hydrate(storage_string)
  }
  as_list() {
    return Item.get_items(this.from_storage())
  }
  as_object() {
    return Item.get_first_item(this.from_storage())
  }
  save() {
    // save the information to local storage and if appropriate to the server
    let items = document.querySelector(this.selector)
    if (!items) { return false }
    items = items.outerHTML
    localStorage.setItem(this.item_type, items)
    if (['person', 'posts'].includes(this.item_type)) {
      this.persist(items)
    }
    return true
  }
  sync() {
    // check if an 'item_type'.html already exist for this user.
    // If so merge it's content whith what's stored locally
    const url = this.get_server_file_url()
    if (['person', 'posts'].includes(this.item_type) && navigator.onLine) {
      firebase.auth().onAuthStateChanged(user => {
        if (user) {
          const doc_u_path = `/people/${user.phoneNumber}/${this.filename}`
          firebase.storage().ref().child(doc_u_path).getDownloadURL().then((url) => {
            fetch(url).then(response => {
              response.text().then(server_text => {
                const server_as_fragment = Storage.hydrate(server_text)
                let server_as_list = Item.get_items(server_as_fragment)
                console.log(server_as_list)
              })
              let items = [...this.from_storage, ...from_server]
              // console.log(items)
            })
          })
        }
      })
    }
  }

  get_server_file_url() {

  }

  persist(doc_u_ment, doc_u_path) {
    firebase.auth().onAuthStateChanged(user => {
      if (user && navigator.onLine) {
        const file = new File([doc_u_ment], this.filename)
        if (!doc_u_path) {
          doc_u_path = `/people/${user.phoneNumber}/${this.filename}`
        }
        firebase.storage().ref()
          .child(doc_u_path)
          .put(file, this.metadata)
          .catch(console.log.bind(console))
      }
    })
  }
}

export default Storage
export const person_storage = new Storage('person')
export const posts_storage = new Storage('posts', '[itemprop=posts]')
export const activity_storage = new Storage('activity', '[itemprop=activity]')
export const relations = new Storage('relations', '[itemprop=relations]')
export const phonebook = new Storage('phonebook')

//  use this for firebase logging
// .then( snapshot => {
//   console.log('Uploaded', snapshot.totalBytes, 'bytes.')
//   console.log(snapshot.metadata)
//   console.log(snapshot.fullPath)
//   console.log('File available at', snapshot.downloadURL)
// })
