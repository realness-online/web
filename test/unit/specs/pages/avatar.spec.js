import { shallow, createLocalVue } from 'vue-test-utils'
import Storage from '@/modules/Storage'
import VueRouter from 'vue-router'
import avatar from '@/pages/avatar'
describe('@/pages/avatar.vue', () => {
  let wrapper
  const $route = {
    params: {}
  }
  beforeEach(() => {
    wrapper = shallow(avatar, {
      mocks: {
        $route
      }
    })
  })
  describe('displaying an avatar', () => {
    it('renders the silhouette by default', () => {
      expect(wrapper.element).toMatchSnapshot()
    })
  })

  it('finished_viewing()', () => {
    const localVue = createLocalVue()
    localVue.use(VueRouter)
    const router = new VueRouter()
    sessionStorage.setItem('previous', '/test_route')
    let wrapper = shallow(avatar, {
      localVue,
      router
    })
    wrapper.vm.finished_viewing()
    expect(wrapper.vm.$route.path).toBe('/test_route')
    sessionStorage.removeItem('previous')
  })
  it('open_camera()', () => {
    wrapper.setProps({view_avatar: true})
    let mock_click = jest.fn()
    wrapper.vm.$refs.file_upload.click = mock_click
    wrapper.vm.open_camera()
    expect(mock_click).toBeCalled()
  })
  describe('accept_changes()', () => {
    it('should update the avatar', () => {
      const localVue = createLocalVue()
      localVue.use(VueRouter)
      const router = new VueRouter()
      let wrapper = shallow(avatar, {
        localVue,
        router
      })
      const save_spy = jest.fn(() => Promise.resolve('save_spy'))
      jest.spyOn(Storage.prototype, 'save').mockImplementation(save_spy)
      wrapper.vm.accept_changes()
      expect(save_spy).toBeCalled()
    })
    it('should save person')

    it('should trigger change event on file input', () => {
      wrapper.setProps({view_avatar: true})
      let input = wrapper.find('input[type=file]')
      expect(input.exists()).toBe(true)
      input.element.value = ''
      input.trigger('change')
      // currently no way to test file inputs. let's trigger the event anyway
    })
  })
})
