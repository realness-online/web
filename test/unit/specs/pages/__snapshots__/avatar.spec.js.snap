// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@/pages/avatar.vue displaying an avatar renders the silhouette by default 1`] = `
<section
  id="upload"
>
  <input
    accept="image/jpeg"
    capture=""
    type="file"
  />
   
  <!---->
   
  <footer>
    <menu>
      <button>
        <!---->
      </button>
       
      <a>
        <!---->
      </a>
       
      <a>
        <!---->
      </a>
    </menu>
  </footer>
</section>
`;
